### PR TITLE
PSP2: Fix vitasdk homebrew loading

### PR DESCRIPTION
**Pull request by frangarcj**
With this PR, vitasdk compiled homebrews are loaded into armv7 module memory. Module info is resolved and relocations are done

#3956

### CODE REVIEW

**Comment by hollow147** [comment]
sem titulo


Is there any way to solve this?

**Comment by clienthax** [comment]
@hollow147 don't spam unrelated PRS with crap please

**Comment by hollow147** [comment]
@clienthax I'm sorry, I'll delete

**Comment by clienthax** [comment]
This seems to break on short relocations with official selfs/elfs, comments in discord

**Comment by JXPZ** [comment]
esboco 4
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1606389/RPCS3.zip)

**Comment by frangarcj** [comment]
@JXPZ could you try again?

**Comment by JXPZ** [comment]
esboco 1

**Comment by frangarcj** [comment]
I think my changes should not affect that but maybe someone can confirm

**Comment by miri1337** [comment]
Works for me.

File: rpcs3/Crypto/unself.cpp [hunk 1]
```diff
@@ -1451,13 +1658,21 @@ static bool CheckDebugSelf(fs::file& s)
 	// Check for DEBUG version.
 	if (key_version == 0x80 || key_version == 0xc0)
 	{
-		LOG_WARNING(LOADER, "Debug SELF detected! Removing fake header...");
+		
+		s.seek(0x04);
+		const u16 version = s.read<le_t<u16>>();
 
+		if (version == 3)
```
[the remark this answers is not in the document]
Comment (frangarcj) [reply rpcs3/Crypto/unself.cpp]: Done

File: rpcs3/Emu/PSP2/ARMv7Module.cpp [hunk 2]
```diff
@@ -314,27 +366,243 @@ void arm_load_exec(const arm_exec_object& elf)
...
 
 			if (!start_addr)
 			{
-				start_addr = prog.p_vaddr;
+				start_addr = p_vaddr[i];
+			}
+
+			std::memcpy(vm::base(p_vaddr[i]), prog.bin.data(), prog.p_filesz);
+		}
+		else if (prog.p_type == 0x60000000) {
```
Comment (kd-11) [inline rpcs3/Emu/PSP2/ARMv7Module.cpp]: brace

File: rpcs3/Emu/PSP2/ARMv7Module.cpp [hunk 3]
```diff
@@ -314,27 +366,243 @@ void arm_load_exec(const arm_exec_object& elf)
...
+					*   J2    = lower[11]
+					*/
+					sign = (upper >> 10) & 1;
+					j1 = (lower >> 13) & 1;
+					j2 = (lower >> 11) & 1;
+					offset = r_addend + symval - loc;
+
+					if (offset <= (s32)0xff000000 ||
+						offset >= (s32)0x01000000) {
+						LOG_NOTICE(LOADER, "reloc %x out of range: 0x%08X", i, symval);
```
Comment (kd-11) [inline rpcs3/Emu/PSP2/ARMv7Module.cpp]: here too

File: rpcs3/Emu/PSP2/ARMv7Module.cpp [hunk 4]
```diff
@@ -314,27 +366,243 @@ void arm_load_exec(const arm_exec_object& elf)
...
+
+					value = ((u32)lower << 16) | upper;
+				}
+				break;
+				case R_ARM_CALL:
+				case R_ARM_JUMP24:
+				{
+					offset = r_addend + symval - loc;
+					if (offset <= (s32)0xfe000000 ||
+						offset >= (s32)0x02000000) {
```
Comment (kd-11) [inline rpcs3/Emu/PSP2/ARMv7Module.cpp]: and here

File: rpcs3/Crypto/unself.cpp [hunk 5]
```diff
@@ -1285,20 +1516,74 @@ fs::file SELFDecrypter::MakeElf(bool isElf32)
...
+					// Seek to the program header data offset and write the data.
+					e.seek(phdr32_arr[meta_shdr[i].program_idx].p_offset);
+					e.write(data_buf.get() + data_buf_offset, meta_shdr[i].data_size);
+
+					// Advance the data buffer offset by data size.
+					data_buf_offset += meta_shdr[i].data_size;
+				}
+			}
+		}
+		else {
```
Comment (kd-11) [inline rpcs3/Crypto/unself.cpp]: missed one

**Comment by Nekotekina** [comment]
Merged to psp2 branch. Thanks for contribution.

**Pull request closed without merging**